### DISCUSSION TITLE
Diagonal operators

### DISCUSSION THREAD

**Discussion by gcarleo** [netket/netket#1942] [Ideas]
Maybe we already discussed this at some point, but it would be nice if we had a subclass of operators that are DiagonalOperator, in such a way that we could easily overload also applying nonlinear functions to these operators f(Op), whereas clearly it is problematic to do so for a general Operator

**Comment by PhilipVinc** [comment]
But is the nonlinear function acting element wise on the elements? or what kind of function? an exponential?

**Reply by gcarleo** [reply]
a general function f that acts on a diagonal operator, so the new matrix elements are diagonal and just f(Mels). also we could implement some specific optimizations for diagonal operators that we don't really do now

**Reply by PhilipVinc** [reply]
Ok, so you mean a function applied element-wise to the operator. 
Yeah it's not a bad idea. We just need to store the function and apply it after we extract the matrix elements.
This is easy to do if someone defines a wrapper operator type.,
Technically we can already check if an operator is diagonal so it can throw an error if the input is not diagonal...
